### PR TITLE
Move SwiftSyntaxParser related code to its own module

### PR DESCRIPTION
**Pull request by rintaro**
* Remove `_SyntaxParserInterop.swift`
* Use `RawSyntax.toOpaque()`/`RawSyntax.fromOpaque(:_)` to interop
* Move `RawSyntax.createFromCSyntaxNode()` to the module
* Move SyntaxKind/RawTokenKind/RawTriviaPiece creation from C serialization code to the module
* Tweak some RawSyntax API visibility
* Refactor materialized token creation API to efficiently create a token from SyntaxParser modules
* Introduce `Syntax.withUnsafeRawSyntax(_:)` API

### CODE REVIEW

File: Sources/SwiftSyntaxParser/SyntaxParser.swift [hunk 1]
```diff
@@ -286,6 +287,3 @@ extension Diagnostic {
   }
 }
 
-fileprivate func verifyCNodeLayoutHash() -> Bool {
-  return cNodeLayoutHash() == SwiftSyntax.cNodeLayoutHash()
```
Comment (rintaro) [inline Sources/SwiftSyntaxParser/SyntaxParser.swift:290]: Now that `SwiftSyntax` doesn't know anything about `CNodes`, so this is not needed anymore.

**Comment by rintaro** [comment]
@swift-ci Please test

File: Sources/SwiftSyntaxParser/RawSyntax+CSwiftSyntax.swift [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    return makeLayout(
+      kind: syntaxKind,
+      uninitializedCount: cnodes.count,
+      arena: arena
+    ) { buffer in
+      guard var ptr = buffer.baseAddress else { return }
+      for cnode in cnodes {
+        let element: RawSyntax? = cnode.map {
+          // 'CClinentNode' is a node created with `toOpaque()`
```
Comment (ahoppen) [inline Sources/SwiftSyntaxParser/RawSyntax+CSwiftSyntax.swift:94]: Typo
```suggestion
          // 'CClientNode' is a node created with `toOpaque()`
```

File: Sources/SwiftSyntaxParser/RawSyntax+CSwiftSyntax.swift [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        initializeRawTriviaBuffer($0, textRange.upperBound, trailingTriviaInfo)
+      })
+  }
+
+  static func makeLayoutFromCNode(_ p: CSyntaxNodePtr, arena: SyntaxArena) -> RawSyntax {
+    let cnode = p.pointee
+    let syntaxKind = SyntaxKind(serializationCode: cnode.kind)
+    assert(syntaxKind != .token)
+
+    let cnodes = UnsafeBufferPointer<CClientNode?>(
```
Comment (ahoppen) [inline Sources/SwiftSyntaxParser/RawSyntax+CSwiftSyntax.swift:83]: Maybe call `cnodes` `children` instead?

**Review by ahoppen — COMMENTED** [review]
Nice. This is great work, making the `SwiftSyntax` module smaller.

**Review by ahoppen — APPROVED** [review]

**Review by CodaFi — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please test

**Pull request merged**